# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/WindowGlobalChild.cpp
# Commit: 6709a3bc1a1c
# Full Hash: 6709a3bc1a1c4ac18dce0f6db78efe12b3e1f178
# Author: Kashav Madan <kmadan@mozilla.com>
# Date: 2019-07-15 21:39:46
# Regressor Bug: 1543251
# File Overlap Count: 1
# Description:
#   Bug 1543251 - Move hasBeforeUnload from PBrowser to PWindowGlobal, r=nika
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D37003
# ==============================================================================

diff -r 83881815f007 -r 6709a3bc1a1c dom/ipc/WindowGlobalChild.cpp
--- a/dom/ipc/WindowGlobalChild.cpp	Mon Jul 15 11:01:04 2019 +0000
+++ b/dom/ipc/WindowGlobalChild.cpp	Mon Jul 15 17:30:26 2019 +0000
@@ -46,6 +46,7 @@
       mBrowsingContext(aBrowsingContext),
       mInnerWindowId(aWindow->WindowID()),
       mOuterWindowId(aWindow->GetOuterWindow()->WindowID()),
+      mBeforeUnloadListeners(0),
       mIPCClosed(true) {}
 
 already_AddRefed<WindowGlobalChild> WindowGlobalChild::Create(
@@ -157,6 +158,26 @@
   return !BrowsingContext()->GetEmbedderElement();
 }
 
+void WindowGlobalChild::BeforeUnloadAdded() {
+  // Don't bother notifying the parent if we don't have an IPC link open.
+  if (mBeforeUnloadListeners == 0 && !mIPCClosed) {
+    SendSetHasBeforeUnload(true);
+  }
+
+  mBeforeUnloadListeners++;
+  MOZ_ASSERT(mBeforeUnloadListeners > 0);
+}
+
+void WindowGlobalChild::BeforeUnloadRemoved() {
+  mBeforeUnloadListeners--;
+  MOZ_ASSERT(mBeforeUnloadListeners >= 0);
+
+  // Don't bother notifying the parent if we don't have an IPC link open.
+  if (mBeforeUnloadListeners == 0 && !mIPCClosed) {
+    SendSetHasBeforeUnload(false);
+  }
+}
+
 void WindowGlobalChild::Destroy() {
   // Perform async IPC shutdown unless we're not in-process, and our
   // BrowserChild is in the process of being destroyed, which will destroy us as